coordinator/internal/stateguard/stateguard.go

```diff
@@ -0,0 +1,445 @@
+		// the meantime. This is fine: we know that m.state must be a transition after ours because
+		// the SetLatest call succeeded. That other UpdateState call must have been operating on our
+		// nextState already, because it had to refer to our transition. Thus, we can ignore the return
+		// value of CompareAndSwap, but we need to return the (now intermediate) state that the caller
+		// intended and we must not unconditionally force our state as current or update the gauge.
+		return nextState, nil
+	}
+	g.metrics.manifestGeneration.Set(float64(nextState.generation))
+	return nextState, nil
+}
+
+// GetHistory returns a list of manifests, the current manifest being last, and the policies
+// referenced in at least one of the manifests.
+func (g *Guard) GetHistory() ([][]byte, map[manifest.HexString][]byte, error) {
+	state, err := g.GetState()
+	if err != nil {
+		return nil, nil, err
+	}
+	var manifests [][]byte
+	policies := make(map[manifest.HexString][]byte)
+	err = g.hist.WalkTransitions(state.latest.TransitionHash, func(_ [history.HashSize]byte, t *history.Transition) error {
+		manifestBytes, err := g.hist.GetManifest(t.ManifestHash)
+		if err != nil {
+			return err
+		}
+		manifests = append(manifests, manifestBytes)
+
+		var mnfst manifest.Manifest
+		if err := json.Unmarshal(manifestBytes, &mnfst); err != nil {
+			return err
+		}
+
+		for policyHashHex := range mnfst.Policies {
+			if _, ok := policies[policyHashHex]; ok {
+				continue
+			}
+			policyHash, err := policyHashHex.Bytes()
+			if err != nil {
+				return fmt.Errorf("converting hex to bytes: %w", err)
+			}
+			var policyHashFixed [history.HashSize]byte
+			copy(policyHashFixed[:], policyHash)
+			policyBytes, err := g.hist.GetPolicy(policyHashFixed)
+			if err != nil {
+				return fmt.Errorf("getting policy: %w", err)
+			}
+			policies[policyHashHex] = policyBytes
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetching manifests from history: %w", err)
+	}
+	// Traversing the history yields manifests in the wrong order, so reverse the slice.
+	slices.Reverse(manifests)
+
+	return manifests, policies, nil
+}
+
+// State is a snapshot of the Coordinator's manifest history.
+type State struct {
+	seedEngine    *seedengine.SeedEngine
+	manifest      *manifest.Manifest
+	manifestBytes []byte
+	ca            *ca.CA
+
+	latest     *history.LatestTransition
+	generation int
+
+	// stale is set to true when we discover that this State is not current anymore.
+	// This field is only ever flipped from false to true!
+	stale atomic.Bool
+}
+
+// NewStateForTest constructs a new State object.
+//
+// This function is intended for testing packages that work on State objects. It fills the fields
+// that are observable outside this package, but does not manage the fields only relevant for this
+// package. State objects created with this function can't be used as arguments to this package's
+// public API functions.
```

Comment:
imo it is bad practice exposing such test related functions as part of the normal package api. It hints at the fact that there should be an interface and stubbing used at the other side.